File: backend/complex/make-payment/internal/interceptors/logging.go
package interceptors

import (
	"context"
	"time"

	"github.com/sirupsen/logrus"
	"google.golang.org/grpc"
)

func LoggingInterceptor(logger *logrus.Logger) grpc.UnaryServerInterceptor {
	return func(
		ctx context.Context,
		req interface{},
		info *grpc.UnaryServerInfo,
		handler grpc.UnaryHandler,
	) (res interface{}, err error) {
		start := time.Now()

		logger.WithFields(logrus.Fields{
			"call_type": "Unary",
			"method":    info.FullMethod,
			"start":     start.Format(time.RFC3339),
		}).Info("request received")

		res, err = handler(ctx, req)

		fields := logrus.Fields{
			"call_type": "Unary",
			"method":    info.FullMethod,
			"took":      time.Since(start).String(),
		}

		if err != nil {
			fields["success"] = false
			logger.WithFields(fields).Error("request processed")
		} else {
			fields["success"] = true
			logger.WithFields(fields).Info("request processed")
		}

		return res, err
	}
}

File: backend/complex/save-notes/src/interceptors/logging.py
import logging
import time

import grpc

class LoggingInterceptor(grpc.ServerInterceptor):
    def intercept_service(self, continuation, handler_call_details):
        start_time = time.time()
        method = handler_call_details.method
        response = continuation(handler_call_details)
        elapsed_time = time.time() - start_time 
        logging.info(f"{method} - Took {elapsed_time} seconds")     

        return response